### PR TITLE
SB-25252 - migration script to populate the score and max score data

### PR DESCRIPTION
**Pull request by manjudr**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-25252" title="SB-25252" target="_blank">Sub-taskSB-25252</a>  Migration script to populate the best_score and max_score into user_activity_agg table
  </td></table>
  <br />
 

Activity Migration Data product will execute only on demand, Which used to migrate the activity_agg table data by reading the best attempt score and max score for the specific batch/course from the `assessment_aggregator` and updates into `user_activity_agg` table

Currently, In the user_activity_agg table `agg` column contains only completed content counts.  This job will populate both `agg` and `agg_last_updated` columns with the best score and max score values.

ex: 
```
agg - Map("score:$content_id"-> $score,  "max_score:$content_id"-> $max_score)

agg_last_updated - Map("score:$content_id"-> $currentTimeStamp,  "max_score:$content_id"-> $currentTimeStamp)

```

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    JobLogger.start(s"$jobName started executing", Option(Map("config" -> config, "model" -> jobName)))
+
+    implicit val jobConfig: JobConfig = JSONUtils.deserialize[JobConfig](config)
+    implicit val spark: SparkSession = openSparkSession(jobConfig)
+    implicit val frameworkContext: FrameworkContext = getReportingFrameworkContext()
+    spark.setCassandraConf("LMSCluster", CassandraConnectorConf.ConnectionHostParam.option(AppConf.getConfig("sunbird.courses.cluster.host")))
+    try {
+      val res = CommonUtil.time(migrateData(spark, jobConfig))
+      val total_records = res._2.count()
+      JobLogger.log(s"Updating the $total_records records in the cassandra table table", None, INFO)
```
Comment (maheshkumargangula) [inline data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala]:
```suggestion
      JobLogger.log(s"Updating the $total_records records in the cassandra table", None, INFO)
```

File: data-products/src/test/scala/org/sunbird/analytics/audit/TestScoreMetricMigrationJob.scala [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  it should "Should migrate the score metrics data" in {
+    implicit val mockFc: FrameworkContext = mock[FrameworkContext]
+    val strConfig = """{"search":{"type":"none"},"model":"org.sunbird.analytics.audit.ScoreMetricMigrationJob","modelParams":{"store":"azure","sparkCassandraConnectionHost":"localhost", "fromDate":"$(date --date yesterday '+%Y-%m-%d')","toDate":"$(date --date yesterday '+%Y-%m-%d')", "metricsType":"attempt_metrics", "forceMerge": false},"parallelization":8,"appName":"ScoreMetricMigrationJob"}""".stripMargin
+    implicit val jobConfig: JobConfig = JSONUtils.deserialize[JobConfig](strConfig)
+    implicit val sc: SparkContext = spark.sparkContext
+    val res = ScoreMetricMigrationJob.migrateData(spark, jobConfig)
+    res.count() should be(6)
+    println("ress" + res.show(false))
```
Comment (maheshkumargangula) [inline data-products/src/test/scala/org/sunbird/analytics/audit/TestScoreMetricMigrationJob.scala]:
```suggestion

```

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

**Review by maheshkumargangula — APPROVED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    val assessmentAggDF = getBestScoreAggDetailsDF(fetchAssessmentData(session, batchIds), metricsType)
+
+    val filterDF = activityAggDF.join(assessmentAggDF, activityAggDF.col("activity_id") === assessmentAggDF.col("course_id") &&
+      activityAggDF.col("userid") === assessmentAggDF.col("user_id") &&
+      activityAggDF.col("context_id") === assessmentAggDF.col("batchid"), joinType = "inner")
+      .select("agg", "aggregates", "agg_last_updated", "activity_type", "user_id", "context_id", "activity_id", "total_max_score", "best_score", "content_id", "agg_details", "migrating_agg_details", "attempts_count")
+
+    filterDF.withColumn("aggregates", updateAggColumn(when(lit(forceMerge), col("agg").cast("map<string, double>")).otherwise(col("aggregates").cast("map<string, double>")), col("total_max_score").cast(sql.types.DoubleType), col("best_score").cast(sql.types.DoubleType), col("attempts_count").cast(sql.types.DoubleType), col("content_id").cast(sql.types.StringType)))
+      .withColumn("agg_details", when(lit(forceMerge), col("migrating_agg_details")).otherwise(updatedAggDetailsCol(col("agg_details"), col("migrating_agg_details"))))
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala:81]: The forceMerge logic can be moved to the UDF. This code would be much simpler to read

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala (hunk 3 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/audit/ScoreMetricMigrationJob.scala:81]: Santhosh, We already have a udf function calling in this code. 
Creating nested udf is creating some issue. Could we please take up this review comment in next release ?

**Review by SanthoshVasabhaktula — APPROVED** [review]

**Pull request merged**